### PR TITLE
6.x to 7.x Porting : Batching refactor - Multipart mime only

### PR DESCRIPTION
**Pull request by manivannanse**
### Issues

*Ported multi part mime refactoring from 6.X.*

### Description

*Refactored code to support multiple batch formats. This supports only support multi part mime*

### Checklist (Uncheck if it is not completed)

- [  ] Test cases added
- [  ] Build and test with one-click build and test script passed

### Additional work necessary
4 Test failures needs to be fixed. Details are as follows
1 - Public Api test. This can be fixed by updating baseline file. But not sure if this is correct
2-  of them are ODataMessageTests.cs refereing to ODatabatch Writter and Reader. With new model it is abstract class. There are test changes for this. But I haven't ported this yet. Need to make sure this is not breaking change before porting test code
1 - renaming ChangeSetStarted to ChangesetStarted broke the test.

### CODE REVIEW

File: src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReaderStream.cs [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+        private void DetermineChangesetBoundaryAndEncoding(string contentType)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(contentType), "Should have validated that non-null, non-empty content type header exists.");
+
+            ODataMediaType mediaType;
+            ODataPayloadKind readerPayloadKind;
+            MediaTypeUtils.GetFormatFromContentType(
+                contentType,
+                new ODataPayloadKind[] { ODataPayloadKind.Batch },
+                new ODataMediaTypeResolver(),
```
Comment (manivannanse) [inline src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReaderStream.cs]: >ODataMediaTypeResolver() [](start = 20, length = 24)

In 6.x it was ODataMediaTypeResolver.DefaultMediaTypeResolver. Need to revisit this

**Comment by manivannanse** [comment]
Test failures still exists are as follows. These can be fixed by porting appropriate test files. Would like to understand why refactoring needs test fixes before porting the test code

13 Test failures needs to be fixed. Details are as follows
9 of them are -  batch Buffer was converted to upper case. Instead of fixing the test i changed the code to fix this. Need to figure out which one is correct

1 - Public Api test. This can be fixed by updating baseline file. But not sure if this is correct
2-  of them are ODataMessageTests.cs refereing to ODatabatch Writter and Reader. With new model it is abstract class. There are test changes for this. But I haven't ported this yet. Need to make sure this is not breaking change before porting test code
1 - renaming ChangeSetStarted to ChangesetStarted broke the test.

File: src/Microsoft.OData.Core/ODataBatchReader.cs [hunk 2]
```diff
@@ -17,24 +17,23 @@ namespace Microsoft.OData
     #endregion Namespaces
 
     /// <summary>
-    /// Class for reading OData batch messages; also verifies the proper sequence of read calls on the reader.
+    /// Abstract Class for reading OData batch messages; also verifies the proper sequence of read calls on the reader.
     /// </summary>
-    public sealed class ODataBatchReader : IODataBatchOperationListener
+    public abstract class ODataBatchReader : IODataBatchOperationListener
```
Comment (manivannanse) [inline src/Microsoft.OData.Core/ODataBatchReader.cs]: >abstract [](start = 11, length = 8)

Is this breaking change.  I see tests that were initializing the reader and writer. Need to analyzed more of usage patterns

File: src/Microsoft.OData.Core/ODataBatchReaderStream.cs [hunk 3]
```diff
@@ -10,55 +10,34 @@ namespace Microsoft.OData
...
-
-        /// <summary>The boundary string for the batch structure itself.</summary>
-        private readonly string batchBoundary;
 
         /// <summary>The buffer used by the batch reader stream to scan for boundary strings.</summary>
-        private readonly ODataBatchReaderStreamBuffer batchBuffer;
-
-        /// <summary>The media type resolver to use when interpreting the incoming content type.</summary>
-        private readonly ODataMediaTypeResolver mediaTypeResolver;
+        protected readonly ODataBatchReaderStreamBuffer batchBuffer;
```
Comment (manivannanse) [inline src/Microsoft.OData.Core/ODataBatchReaderStream.cs]: >batchBuffer [](start = 56, length = 11)

This is renamed to BatchBuffer in original pull request in 6.x branch. This casued 6 test failure. Theoretically refactoring should not break any existing tests.

File: src/Microsoft.OData.Core/ODataBatchWriter.cs [hunk 4]
```diff
@@ -899,14 +688,14 @@ private void ValidateTransition(BatchWriterState newState)
...
                         throw new ODataException(Strings.ODataBatchWriter_InvalidTransitionFromBatchStarted);
                     }
 
                     break;
-                case BatchWriterState.ChangeSetStarted:
-                    if (newState != BatchWriterState.OperationCreated && newState != BatchWriterState.ChangeSetCompleted)
+                case BatchWriterState.ChangesetStarted:
+                    if (newState != BatchWriterState.OperationCreated && newState != BatchWriterState.ChangesetCompleted)
                     {
                         throw new ODataException(Strings.ODataBatchWriter_InvalidTransitionFromChangeSetStarted);
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/ODataBatchWriter.cs]: It looks like we decided to change the letter casing for "Change**S**etStarted" to "Change**s**etStarted". We should probably do a search across the project for other variables that have "ChangeSet" and update them for consistency.

ODataBatchWriter_InvalidTransitionFromChange**S**etStarted

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataBatchReaderStreamTests.cs [hunk 5]
```diff
@@ -10,7 +10,7 @@
 using FluentAssertions;
 using Xunit;
 using ErrorStrings = Microsoft.OData.Strings;
-
+using Microsoft.OData.MultipartMixed;
```
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataBatchReaderStreamTests.cs]: nit: empty line after this line

File: src/Microsoft.OData.Core/ODataBatchReader.cs [hunk 6]
```diff
@@ -17,24 +17,23 @@ namespace Microsoft.OData
...
     /// <summary>
-    /// Class for reading OData batch messages; also verifies the proper sequence of read calls on the reader.
+    /// Abstract Class for reading OData batch messages; also verifies the proper sequence of read calls on the reader.
     /// </summary>
-    public sealed class ODataBatchReader : IODataBatchOperationListener
+    public abstract class ODataBatchReader : IODataBatchOperationListener
     {
         /// <summary>The input context to read the content from.</summary>
-        private readonly ODataRawInputContext inputContext;
+        private  ODataInputContext inputContext;
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/ODataBatchReader.cs]: nit: there's an extra space between "private" and "ODataInputContext"

**Comment by robward-ms** [comment]
If we want to accept this portion of batching with failing tests, I suggest we target a feature branch instead of submitting to master.

File: src/Microsoft.OData.Core/ODataBatchReaderStream.cs (hunk 3 above)
Comment (AlanWong-MS) [reply src/Microsoft.OData.Core/ODataBatchReaderStream.cs]: I saw four failures in the build. Did you end up fixing some?

File: src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReader.cs [hunk 7]
```diff
@@ -0,0 +1,472 @@
+//---------------------------------------------------------------------
+// <copyright file="ODataMultipartMixedBatchReader.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.MultipartMixed
```
Comment (mtroth-microsoft) [inline src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReader.cs]: >MultipartMixed [](start = 26, length = 14)

not enough classes here to merit its own namespace.

File: src/Microsoft.OData.Core/ODataBatchReader.cs [hunk 8]
```diff
@@ -204,440 +236,81 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
         }
 
         /// <summary>
-        /// Returns the next state of the batch reader after an end boundary has been found.
+        /// Increases the size of the current batch message; throws if the allowed limit is exceeded.
         /// </summary>
-        /// <returns>The next state of the batch reader.</returns>
-        private ODataBatchReaderState GetEndBoundaryState()
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/ODataBatchReader.cs]: >GetEndBoundaryState [](start = 38, length = 19)

I would expect these states and transitions to be common between the multipart/mixed and JSON formats, in which case they should be in the base class.

File: src/Microsoft.OData.Core/ODataBatchWriter.cs [hunk 9]
```diff
@@ -136,66 +110,92 @@ private enum BatchWriterState
...
+                {
+                    Debug.Assert(!this.outputContext.WritingResponse, "Request message can only be set when writing request.");
+                    return this.currentOperationRequestMessage.OperationMessage;
+                }
+                else if (this.currentOperationResponseMessage != null)
+                {
+                    Debug.Assert(this.outputContext.WritingResponse, "Response message can only be set when writing response.");
+                    return this.currentOperationResponseMessage.OperationMessage;
+                }
+                else
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/ODataBatchWriter.cs]: >               else [](start = 1, length = 19)

Don't really need the "else" as both the previous cases end in returns.

File: src/Microsoft.OData.Core/ODataBatchWriter.cs [hunk 10]
```diff
@@ -136,66 +110,92 @@ private enum BatchWriterState
...
+            set { this.currentOperationContentId = value; }
+        }
+
+        /// <summary>
+        /// Gets or Sets the batch writer's state.
+        /// </summary>
+        protected BatchWriterState State
+        {
+            get { return this.state; }
+            set { this.state = value; }
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/ODataBatchWriter.cs]: > set { this.state = value; } [](start = 11, length = 28)

I would generally expect the state to be set within the ODataBatchWriter base class as the derived classed called methods for each transition.

File: src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReader.cs [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+
+                case ODataBatchReaderState.Exception:    // fall through
+                case ODataBatchReaderState.Completed:
+                    Debug.Assert(false, "Should have checked in VerifyCanRead that we are not in one of these states.");
+                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_ReadImplementation));
+
+                default:
+                    Debug.Assert(false, "Unsupported reader state " + this.State + " detected.");
+                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_ReadImplementation));
+            }
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReader.cs]: Ideally state transitions would be done in the base class; the derived class would never set the state directly, but would call the next "Startxyz" method in the base class, which would transition state and then call the implementation method in the derived class.  Take a look at how this is done in ODataReader.

**Comment by AlanWong-MS** [comment]
Closing this PR as it is superseded by #960.

**Pull request closed without merging**